leftwm-core/src/handlers/command_handler/scratchpad_handler.rs

```diff
@@ -0,0 +1,1242 @@
+    scratchpad_window: &WindowHandle,
+) -> Result<(), &'static str> {
+    log::trace!("Show scratchpad window {:?}", scratchpad_window);
+    let current_tag = &manager
+        .state
+        .focus_manager
+        .tag(0)
+        .ok_or("Could not retrieve the current tag")?;
+    let window = manager
+        .state
+        .windows
+        .iter_mut()
+        .find(|w| w.handle == *scratchpad_window)
+        .ok_or("Could not find window from scratchpad_window")?;
+    let previous_tag = window.tag;
+    window.untag();
+
+    // Remove the entry for the previous tag to prevent the scratchpad being
+    // refocused.
+    if let Some(previous_tag) = previous_tag {
+        manager
+            .state
+            .focus_manager
+            .tags_last_window
+            .remove(&previous_tag);
+    }
+    // Show the scratchpad.
+    window.tag(current_tag);
+    window.set_visible(true);
+
+    // send tag changement to X
+    let act = DisplayAction::SetWindowTag(*scratchpad_window, window.tag);
+    manager.state.actions.push_back(act);
+    manager.state.sort_windows();
+    manager.state.handle_window_focus(scratchpad_window);
+    manager.state.move_to_top(scratchpad_window);
+
+    Ok(())
+}
+
+/// With the introduction of `VecDeque` for scratchpads, it is possible that a window gets destroyed
+/// in the middle of the `VecDeque`. This is an abstraction to retrieve the next valid pid from a
+/// scratchpad. While walking the scratchpad windows, invalid pids will get removed.
+fn next_valid_scratchpad_pid(
+    scratchpad_windows: &mut VecDeque<u32>,
+    managed_windows: &[Window],
+) -> Option<u32> {
+    while let Some(window) = scratchpad_windows.pop_front() {
+        if managed_windows.iter().any(|w| w.pid == Some(window)) {
+            scratchpad_windows.push_front(window);
+            return Some(window);
+        }
+
+        log::info!(
+            "Dead window in scratchpad found, discard: window PID: {}",
+            window
+        );
+    }
+
+    None
+}
+
+/// Inverse of `next_valid_pid_test`
+fn prev_valid_scratchpad_pid(
+    scratchpad_windows: &mut VecDeque<u32>,
+    managed_windows: &[Window],
+) -> Option<u32> {
+    while let Some(window) = scratchpad_windows.pop_back() {
+        if managed_windows.iter().any(|w| w.pid == Some(window)) {
+            scratchpad_windows.push_back(window);
+            return Some(window);
+        }
+
+        log::info!(
+            "Dead window in scratchpad found, discard: window PID: {}",
+            window
+        );
+    }
+
+    None
+}
+
+/// Check if the scratchpad is visible on the current tag.
+/// If the name of the scratchpad is incorrect, false will be returned
+fn is_scratchpad_visible<C: Config, SERVER: DisplayServer>(
+    manager: &Manager<C, SERVER>,
+    scratchpad_name: &str,
+) -> bool {
+    let current_tag = if let Some(tag) = manager.state.focus_manager.tag(0) {
+        tag
+    } else {
+        return false;
+    };
```

Comment:
This isn't necessary, but a nifty macro could be made for these two as the same thing is being done